snapmgr: periodically remove unneeded base snaps

dilyn-corner opened the pull request:
Snaps have two types of dependencies: hard dependency and soft dependency.

* A hard dependency is the most salient: the snap declares a `base:` and that base snap must be installed on the system for that snap to be installed.

* A soft dependency is a bit looser: primarily these dependencies are things which provide a content interface for arbitrary snaps to leverage in their execution. This dependency is made stronger if a snap declares a content provider as _the_ `default-provider` for that interface. In such a case, that snap will always be installed as a prerequisite.

In cases where a user installs a snap which has a dependency, it's feasible that the user does not care about whether or not that dependency is installed. Specifically, a user may not care if `core20` is installed, just that `firefox` works when it is. In cases where users remove the snap they care about, they may forget about this dependency relationship and forget to remove that now orphaned snap.

This PR aims to clean up so-called orphans on a periodic basis utilizing `Ensure()`. 

We accomplish this by flagging any snap installed as a dependency of some other snap as being `implicitly-installed: true` and all others as `implicitly-installed: false`.
Just in case a snap is marked as `implicitly-installed: true`, a new task called `orphan-removals` is spawned to remove that snap.

**NOTE**: currently, if a snap is not of `type: base`, it won't be removed by this functionality. (line 1185 of `snapmgr.go` in this PR). The reason for this is explained in the comment on that check: the work of identifying if a non-base dependency is being _used_ still has to be done.

Otherewise, everything tends to work as expected.

SNAPDENG-37045

dilyn-corner committed:
snapstate/flags: add a new flag for snap states

ImplicitlyInstalled will be set to true just in case the snap being
installed as a result of being needed by some other snap, as with
content (default) providers, base snaps.

If a user specifically installs a snap, it is to be marked as explicitly
installed, which means ImplicitlyInstalled is false.

Signed-off-by: Joseph Brock <joseph.brock@canonical.com>
Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapmgr: periodically remove unneeded snaps

ensureDependencyRemoval is an addition to the Ensure functionality which
will iterate through a list of snaps and, given some conditions, will
generate an `orphan-removals` task for them.

The two conditions:
1) The snap is not required by any other snap
   This condition obtains just in case that particular snap *could*
     be removed
2) The snap is implicitly installed
   This condition obtains just in case the ImplicitlyInstalled flag in
     the snap's state is set to true

Signed-off-by: Joseph Brock <joseph.brock@canonical.com>
Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapstate_remove_test: add tests for ensureDependencyRemoval

When ensureDependencyRemoval is called, a particular collection of tasks being added is expected. Check for those tasks.

Signed-off-by: Joseph Brock <joseph.brock@canonical.com>
Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapshotstate: fix defer unlocking for some tests

Some tests will defer unlocking while also unlocking the state before
running some function like Settle. If one of the Managers spawned by
Settle fails, the test failure will indicate that an unlocked mutex
failed to be unlocked instead of reporting the actual error emitted by
the failing Manager.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

Commit:
overlord/snapstate: fix propagation of ImplicitlyInstalled flag from SnapSetup to SnapState during installation

Signed-off-by: Joseph Brock <joseph.brock@canonical.com>

dilyn-corner commented:
@kubiko @alexclewontin :)

dilyn-corner committed:
tests: add tests for implicitly-installed flag

These tests check that the correct value for implicitly-installed is set
for snaps in some simple cases.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner commented on overlord/snapstate/handlers.go:
```diff
@@ -409,6 +409,7 @@ func (m *SnapManager) installOneBaseOrRequired(t *state.Task, snapName string, c
 		RequireTypeBase: requireTypeBase,
 		Transaction:     flags.Transaction,
 		Lane:            flags.Lane,
+		ImplicitlyInstalled: true,
```
Made this change in several other places as well.

dilyn-corner committed:
snapmgr: only create orphan removal task when nonempty

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapmgr: avoid repeated parsing of the state

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapmgr: log canRemove error as debug output

It isn't fatal but worth flagging.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
overlord/snapstate: make createDependencyRemovalTasks private

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

miguelpires commented on tests/main/implicit-dependency-flag-set/task.yaml:
```diff
@@ -0,0 +1,15 @@
+  jq -r '.data.snaps.core18["implicitly-installed"]'  /var/lib/snapd/state.json | MATCH true
+  jq -r '.data.snaps.basic18["implicitly-installed"]' /var/lib/snapd/state.json | MATCH false
+
+cleanup: |
```
`cleanup` should be `restore` I think. IIRC the static checks also enforce a particular order between these sections, so I think it might have to be placed before `execute`. But I may be misremembering, you can run `./run-check --static` to confirm

```suggestion
restore: |
```

dilyn-corner replied:
Seems like. Thanks!

miguelpires commented on tests/main/implicit-dependency-flag-not-changed/task.yaml:
The spread tests are missing a `details:` section with a longer explanation (also required by the static checks)

dilyn-corner replied:
Added, thanks!

miguelpires commented on overlord/snapstate/handlers.go:
This is duplicating all the stuff in `handlers_prereq.go` https://github.com/canonical/snapd/blob/master/overlord/snapstate/handlers_prereq.go I think it needs to be rebased on master

dilyn-corner replied:
Yeah, I noticed that something horrible happened in my rebase. Fixed, pushing soon.

miguelpires reviewed (commented):
Thanks for this. This needs a rebase on master. It doesn't build because it's duplicating the code in https://github.com/canonical/snapd/blob/master/overlord/snapstate/handlers_prereq.go which was moved out of `handlers.go` recently-ish

dilyn-corner committed:
tests: add details for implicit dependency tests

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
tests: cleanup -> restore

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
tests: adjust order (restore->execute)

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

miguelpires commented on overlord/snapstate/snapmgr.go:
```diff
@@ -1692,6 +1692,107 @@ func (m *SnapManager) ensureStoreDownloadsCacheCleaned() error {
+				continue
+			}
+			return nil, nil, err
+		}
```
This is iterating on installed snaps so this should be impossible

dilyn-corner replied:
Fair point, I believe I was confused on what the exact behavior or state at the moment this is checked might be.

miguelpires commented on overlord/snapstate/snapmgr.go:
```diff
@@ -1692,6 +1692,107 @@ func (m *SnapManager) ensureStoreDownloadsCacheCleaned() error {
+				continue
+			}
+			return nil, nil, err
+		}
```
This is reading the global snaps' state repeatedly. The call to `InstalledSnaps` reads it to compute the installed snap list and then, for each one of those, we read the entire state with `Get` and extract the individual SnapStates.

You can use `All` which returns a map of SnapState you could then iterate on

dilyn-corner replied:
Brilliant. No idea how I hadn't seen `All`.

miguelpires commented on overlord/snapstate/snapmgr.go:
```diff
@@ -1692,6 +1692,107 @@ func (m *SnapManager) ensureStoreDownloadsCacheCleaned() error {
+		snapInfo, err := Info(m.state, installedSnap.SnapName(), installedSnap.Revision)
+		if err != nil {
+			return nil, nil, err
+		}
```
snapstate has a simpler `CurrentInfo()` method that does this

miguelpires replied:
Since getting the snap.Info requires a disk read, we may as well move it after the `snapType != snap.TypeBase` check to avoid doing it unnecessarily

miguelpires commented on overlord/snapstate/snapmgr.go:
```diff
@@ -1692,6 +1692,107 @@ func (m *SnapManager) ensureStoreDownloadsCacheCleaned() error {
 }
 
+func createDependencyRemovalTasks(m *SnapManager) ([]string, []*state.TaskSet, error) {
+	if changeInFlight(m.state) {
```
We may also have to add this change to the conflict checks. Otherwise, an install may race with this. If it comes in after this check but before the required base is removed, it might see it as already installed but then it would get removed by this change

dilyn-corner replied:
Makes sense to me; you mean to add a check in `handlers_prereq.go`?

miguelpires reviewed (changes requested):
Thank you for the tweaks. There's at least one panicking test in `snapstate/`.

Left one question around conflict checking that needs to be investigated. Also, it seems there are quite a lot of full snap state reads we can do away with. Let me know if you have thoughts/questions about that

dilyn-corner committed:
snapmgr: only read the snapstate once

This should dramatically reduce independent reads of the state in
the loop, the state should be consistent at this point while this is
happening.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapmgr: move Info check after Type check

Avoid unecessary disk read.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
handlers_prereq: guard prereq fetching against in-flight dependency removal tasks

Avoid a delicate race where some prerequisite snap is slated for
auto-removal but is required by some to-be-installed snap.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
overlord: resolve spread test failures

ensureDependencyRemoval can potentially run very early, before a Device
Context is available. This disrupts some tests.

Additionally, add a tracing logger.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

miguelpires commented:
Hey @dilyn-corner, thanks for the changes. This surfaced a race condition around removal of unused bases and operations that can make snaps use those previously unused bases (install, refreshes, reverts, etc). It's not caused by this PR and it's not difficult to even trigger manually, so it makes sense to fix separately. I've opened https://github.com/canonical/snapd/pull/17541 to address it. Once that's done, we can rebase this on master and we shouldn't need to add any new conflict checking here (assuming this uses the export snapstate functions which it currently does). I'll update this once that's merged
